main.go

```diff
@@ -150,27 +195,184 @@ func readConfig(filename string) (*config, error) {
 	return &c, nil
 }
 
+// myClain define token struct
 type myClaim struct {
 	jwt.Claims
 	Roles []string `json:"roles,omitempty"`
 }
 
-func validateClaims(t string, sufficientRoles []string) error {
+// validateClaims check role fom config sufficetRoles is in token roles claim
+func validateClaims(t string, sufficientRoles []string, username string, addGroup bool, addmembership bool) error {
 	token, err := jwt.ParseSigned(t)
 	if err != nil {
 		return fmt.Errorf("error parsing token: %w", err)
 	}
 
 	claims := myClaim{}
 	if err := token.UnsafeClaimsWithoutVerification(&claims); err != nil {
-		return fmt.Errorf("unable to extract claims from token:%w", err)
+		return fmt.Errorf("unable to extract claims from token: %w", err)
 	}
 	for _, role := range claims.Roles {
+
+		if addGroup {
+			createGroup(role, username)
+		}
+
+		if addmembership {
+			addUserToGroup(role, username)
+		}
+
 		for _, sr := range sufficientRoles {
 			if role == sr {
+				log.Print("validateClaims access granted role " + role + " is in token")
 				return nil
 			}
 		}
 	}
-	return fmt.Errorf("role:%s not found", sufficientRoles)
+	return fmt.Errorf("role: %s not found", sufficientRoles)
+}
+
+// createUser this create user is not exsits
+func createUser(username string) {
+	_, err := user.LookupUser(username)
+
+	var pathError *user.NoFoundError
+	// if no user then add one
+	if ok := errors.As(err, &pathError); ok {
+
+		cmd := exec.Command("usr/sbin/useradd", "-m", "-s", "/bin/bash", "-c", app, username)
+		stdoutStderr, err := cmd.CombinedOutput()
+		if err != nil {
+			log.Printf("user already exists: %s ", err.Error())
+		}
+
+		log.Printf("%s", stdoutStderr)
+	} else {
+		log.Printf("user already exists: %s skip create", username)
+	}
+}
+
+// getLastLogin for a user
+func getLastLogin(username string) time.Time {
+
+	cmd := exec.Command("last", "-1", username)
+	stdoutStderr, err := cmd.CombinedOutput()
+	if err != nil {
+		log.Printf("No lastlogin: %s ", err.Error())
+	}
+
+	s := string(stdoutStderr)
+	split := strings.Split(s, " ")
+	x := strings.TrimSpace(split[2] + " " + split[3] + " " + split[4] + " " + split[5] + " " + split[6] + " " + split[7])
+	return parseloginTime(x)
+}
+
+// parseloginTime get time from pased last command
+func parseloginTime(currenttime string) time.Time {
+
+	layout := time.ANSIC
+	t, err := time.Parse(layout, currenttime)
+
+	if err != nil {
```

Comment:
If parsing of time does not work, nil is returned, this is not good

Reply:
if time parse go wrong i use the current time  to do nothing is better